<launch>
  <arg name="start_delay" default="10.0" />

  <arg name="device_type_camera1"    		default="t265"/>
  <arg name="device_type_camera2"    		default="d4.5"/>		<!-- Note: using regular expression. match D435, D435i, D415... -->
  <arg name="serial_no_camera1"    			default=""/>
  <arg name="serial_no_camera2"    			default=""/>
  <arg name="camera1"              			default="t265"/>
  <arg name="camera2"              			default="d400"/>
  <arg name="tf_prefix_camera1"         default="$(arg camera1)"/>
  <arg name="tf_prefix_camera2"         default="$(arg camera2)"/>
  <arg name="initial_reset"             default="false"/>
  <arg name="reconnect_timeout"         default="6.0"/>
  <arg name="enable_fisheye"            default="true"/>
  <arg name="color_width"               default="848"/>
  <arg name="color_height"              default="480"/>
  <arg name="depth_width"               default="848"/>
  <arg name="depth_height"              default="480"/>
  <arg name="clip_distance"             default="-2"/>
  <arg name="topic_odom_in"             default="odom_in"/>
  <arg name="calib_odom_file"           default=""/>
  <arg name="depth_fps"                 default="30"/>
  <arg name="enable_sync"               default="true"/>

  <group ns="$(arg camera1)">
    <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
      <arg name="device_type"           value="$(arg device_type_camera1)"/>
      <arg name="serial_no"             value="$(arg serial_no_camera1)"/>
      <arg name="tf_prefix"         		value="$(arg tf_prefix_camera1)"/>
      <arg name="initial_reset"         value="$(arg initial_reset)"/>
      <arg name="reconnect_timeout"     value="$(arg reconnect_timeout)"/>
      <arg name="enable_fisheye1"       value="$(arg enable_fisheye)"/>
      <arg name="enable_fisheye2"       value="$(arg enable_fisheye)"/>
      <arg name="topic_odom_in"         value="$(arg topic_odom_in)"/>
      <arg name="calib_odom_file"       value="$(arg calib_odom_file)"/>
      <arg name="enable_pose"           value="true"/>
      <arg name="enable_sync"           value="$(arg enable_sync)"/>
    </include>
  </group>

  <group ns="$(arg camera2)">
    <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
      <arg name="device_type"           value="$(arg device_type_camera2)"/>
      <arg name="depth_fps"             value="$(arg depth_fps)"/>
      <arg name="serial_no"             value="$(arg serial_no_camera2)"/>
      <arg name="tf_prefix"		          value="$(arg tf_prefix_camera2)"/>
      <arg name="initial_reset"         value="$(arg initial_reset)"/>
      <arg name="reconnect_timeout"     value="$(arg reconnect_timeout)"/>
      <arg name="align_depth"           value="true"/>
      <arg name="filters"               value="pointcloud"/>
      <arg name="color_width"           value="$(arg color_width)"/>
      <arg name="color_height"          value="$(arg color_height)"/>
      <arg name="depth_width"           value="$(arg depth_width)"/>
      <arg name="depth_height"          value="$(arg depth_height)"/>
      <arg name="clip_distance"         value="$(arg clip_distance)"/>
      <arg name="enable_sync"           value="$(arg enable_sync)"/>
    </include>
  </group>
  <node type="rviz" name="rviz" pkg="rviz" args="-d $(find pointcloud_publisher)/rviz/3d_mapping_rviz.rviz" />
  <node pkg="tf" type="static_transform_publisher" name="t265_to_d400" args=" -0.009375589 0.015890727 0.028273059 0 0 0 /$(arg tf_prefix_camera1)_link /$(arg tf_prefix_camera2)_link 100"/>
  <node pkg="tf" type="static_transform_publisher" name="map_to_aruco" args="  0.75 0 1.15 3.1415926536 0 1.5707963268   map aruco 100"/>
  <node pkg="pointcloud_publisher" type="marker_map_reference.py" name="marker_map_reference" output="screen"/>
  <node pkg="pointcloud_publisher" type="map_to_odom_transformation.py" name="map_to_odom_transformation" output="screen"/>
  <!--Start RTAB-Map-->
   <include file = "$(find rtabmap_ros)/launch/rtabmap.launch" >
      <arg name = "rtabmap_args" value ="--delete_db_on_start --Mem/UseOdomGravity true --Optimizer/GravitySigma 0.3"/>
      <arg name = "odom_topic" value ="/t265/odom/sample"/>
      <arg name = "frame_id" value ="t265_link"/>
      <arg name = "rgbd_sync" value ="true"/>
      <arg name = "depth_topic" value ="/d400/aligned_depth_to_color/image_raw"/>
      <arg name = "rgb_topic" value ="/d400/color/image_raw"/>
      <arg name = "camera_info_topic" value ="/d400/color/camera_info"/>
      <arg name = "approx_rgbd_sync" value ="false"/>
      <arg name = "visual_odometry" value ="false"/>
       <arg name="publish_tf_map"   value="false"/>
      <param name="cloud_noise_filtering_radius" value="0.05"/>
      <param name="cloud_noise_filtering_min_neighbors" value="2"/>
  </include> 

</launch>
